### PR TITLE
The first draft

### LINKED ISSUE #2278: Incorrectly reports succesfull import when wrong for_each key is used.
### OpenTofu Version

```shell
OpenTofu v1.8.6
on darwin_arm64
+ provider registry.opentofu.org/hashicorp/random v3.6.3
```

### OpenTofu Configuration Files

```hcl
locals {
  items = toset(["a", "b", "c"])
}

resource "random_string" "this" {
  for_each = local.items

  length   = 4
}
```

### Debug Output

```
$ tofu import 'random_string.dynamic["a"]' aaaa
random_string.dynamic["a"]: Importing from ID "aaaa"...
random_string.dynamic["a"]: Import prepared!
  Prepared random_string for import
random_string.dynamic["a"]: Refreshing state... [id=aaaa]

Import successful!

The resources that were imported are shown above. These resources are now in
your OpenTofu state and will henceforth be managed by OpenTofu.

$ tofu import 'random_string.dynamic["d"]' aaaa

Import successful!

The resources that were imported are shown above. These resources are now in
your OpenTofu state and will henceforth be managed by OpenTofu.

$ tofu state ls
random_string.dynamic["a"]
$ 
```

### Expected Behavior

The command `tofu import 'random_string.dynamic["d"]' aaaa` should report an error as no resources are actually imported.

The behaviour is correct when trying to import to a static resource:
```
$ tofu import 'random_string.static' aaaa
Error: resource address "random_string.static" does not exist in the configuration.

Before importing this resource, please create its configuration in the root module. For example:

resource "random_string" "static" {
  # (resource arguments)
}

$ 
```

### Actual Behavior

A successful import is reported.

### Steps to Reproduce

1. tofu init
2. tofu apply

### Additional Context

If I recall correctly, the behaviour in terraform was as expected.

### References

_No response_

### PR DESCRIPTION
**Pull request by Gogotchuri**
I have mostly figured out how to handle multiple resource validation in the case of for_each.
I will add additional tests for it, and nested module key resolution, and use a TDD approach to extend this validation for modules if needed.

*This is a draft PR, and the solution is just a working prototype requiring a lot of cleaning up.*

Resolves #2278

## Checklist

- [ ] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [ ] I have not used an AI coding assistant to create this PR.
- [ ] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.
- [ ] I (and other contributors to this PR) have not looked at the Terraform source code while implementing this PR.

### Go checklist

 

- [ ] I have run golangci-lint on my change and receive no errors relevant to my code.
- [ ] I have run existing tests to ensure my code doesn't break anything.
- [ ] I have added tests for all relevant use cases of my code, and those tests are passing.
- [ ] I have only exported functions, variables and structs that should be used from other packages.
- [ ] I have added meaningful comments to all exported functions, variables, and structs.

### Website/documentation checklist

- [ ] I have locally started the website as [described here](https://github.com/opentofu/opentofu/blob/main/website/README.md) and checked my changes.

### CODE REVIEW

File: internal/tofu/node_resource_plan.go [hunk 1]
```diff
@@ -208,6 +217,38 @@ func (n *nodeExpandPlannableResource) DynamicExpand(evalCtx EvalContext) (*Graph
...
+	var diags tfdiags.Diagnostics
+
+	for _, importTarget := range n.importTargets {
+		if !importTarget.IsFromImportCommandLine() {
+			continue
+		}
+
+		cliTarget := importTarget.CommandLineImportTarget
+		key := cliTarget.Addr.Resource.Key
+		// We don't need to check if the resource address doesn't include a key
```
Comment (cam72cam) [inline internal/tofu/node_resource_plan.go]: How does this work with keyed modules?

File: internal/command/testdata/import-non-existent-key/main.tf [hunk 2]
```diff
@@ -0,0 +1,12 @@
+locals {
+  items = toset(["a", "b"])
+}
+
+resource "test_instance" "this" {
+  for_each = local.items
+}
+
+import {
```
Comment (cam72cam) [inline internal/command/testdata/import-non-existent-key/main.tf]: This is very confusing, why have an import block when testing the CLI only import paths?

Comment (Gogotchuri) [reply internal/command/testdata/import-non-existent-key/main.tf]: Oh, my bad, I was testing the import block, and I have temporarily put it there. Forgot to remove it before pushing.

File: internal/tofu/node_resource_plan.go (hunk 1 above)
Comment (Gogotchuri) [reply internal/tofu/node_resource_plan.go]: I will add a test case for modules too, just to be sure. Additionally, I don't think we have test coverage of importing into "keyed" modules. So it would be nice.
Currently, this checks if the resource we are targeting is keyed; non-keyed resource validations are done separately, regardless of the modules being keyed, and since at that point we have fully expanded instances to check against, that shouldn't be an issue.

File: internal/command/testdata/import-non-existent-key/main.tf (hunk 2 above)
Comment (Gogotchuri) [reply internal/command/testdata/import-non-existent-key/main.tf]: Removed in https://github.com/opentofu/opentofu/pull/3106/commits/3c8052ec4b7c596c098eeaafb3f94e3748a27a97

File: internal/tofu/node_resource_plan.go [hunk 3]
```diff
@@ -194,6 +194,15 @@ func (n *nodeExpandPlannableResource) DynamicExpand(evalCtx EvalContext) (*Graph
 		return nil, diags.ErrWithWarnings()
 	}
 
+	// Import target validation, when the target address contains a for_each key,
```
Comment (diofeher) [inline internal/tofu/node_resource_plan.go]: Hey Ilia, first of all, thanks for doing this work! It was easy to understand what was supposed to be done and I think there are good test cases covering it.

With that being said, I would like to propose an alternate implementation where we reuse more of the existing logic:

1) 
First, we're using `postPlanValidateImports` to validate these imports when we are using `import blocks`. We could reuse that on the `context_import`, after the `walker` is done.

This is similar to what we're using at `internal/tofu/context_plan.go/planWalk` function. `postPlanValidateImports` could be renamed to `postExpansionValidateImports` since only the `importResolver` is being used, it's not using something specific from the `plan`.

2)
Besides that, you would need to add the command cli import target to the `imports` list of `importResolver`, by adding them at the `DynamicExpand` from `nodeExpandPlannableResource`.

Here's the full diff of the prototype:

``` 
diff --git a/internal/tofu/context_import.go b/internal/tofu/context_import.go
index 61ef57db31..6ccc8800e3 100644
--- a/internal/tofu/context_import.go
+++ b/internal/tofu/context_import.go
@@ -331,6 +331,13 @@ func (c *Context) Import(ctx context.Context, config *configs.Config, prevRunSta
 		return state, diags
 	}
 
+	allInsts := walker.InstanceExpander.AllInstances()
+
+	importValidateDiags := c.postPlanValidateImports(walker.ImportResolver, allInsts)
+	if importValidateDiags.HasErrors() {
+		return nil, importValidateDiags
+	}
+
 	// Data sources which could not be read during the import plan will be
 	// unknown. We need to strip those objects out so that the state can be
 	// serialized.
diff --git a/internal/tofu/node_resource_plan.go b/internal/tofu/node_resource_plan.go
index b9ad1947e6..905325813d 100644
--- a/internal/tofu/node_resource_plan.go
+++ b/internal/tofu/node_resource_plan.go
@@ -170,6 +170,15 @@ func (n *nodeExpandPlannableResource) DynamicExpand(evalCtx EvalContext) (*Graph
 		// resolve the import as it's already in the resolved form
 		// In addition, if PreDestroyRefresh is true, we know we are running as part of a refresh plan, immediately before a destroy
 		// plan. In the destroy plan mode, import blocks are not relevant, that's why we skip resolving imports
+		if importTarget.IsFromImportCommandLine() {
+			importAddress := importTarget.CommandLineImportTarget.Addr
+			importResolver.imports[importAddress.String()] = EvaluatedConfigImportTarget{
+				Config: nil,
+				Addr:   importAddress,
+				ID:     importTarget.CommandLineImportTarget.ID,
+			}
+			continue
+		}
 		skipImports := importTarget.IsFromImportBlock() && !n.preDestroyRefresh
 		if skipImports {
 			err := importResolver.ExpandAndResolveImport(context.TODO(), importTarget, evalCtx)
```

We reuse existing validations so there's less code to maintain and less branches.

Comment (Gogotchuri) [reply internal/tofu/node_resource_plan.go]: Thanks, Dio, for the extensive write-up and the prototype. I appreciate that!
I like reusing existing validation of import blocks for the CLI block; that's an excellent solution. 
The only reason I initially discarded this approach was to keep the change as minimal as possible. Mixing CLI import and import blocks, in my mind, would cause larger refactors for the edge cases, where we are using config from the import block.
Now that I look at your prototype, this might work, with minimal addition, and in the long run, this is a better way to do it. 
I will add a test case for importing into expanding modules and see if I can use this solution!

Comment (Gogotchuri) [reply internal/tofu/node_resource_plan.go]: The approach worked well. See it implemented in https://github.com/opentofu/opentofu/pull/3106/commits/ba37553d8b1de5a6835b781a3dfd06a096721f2b.
Thanks, Dio!

File: internal/tofu/node_resource_plan.go (hunk 1 above)
Comment (Gogotchuri) [reply internal/tofu/node_resource_plan.go]: I am resolving this thread for now, since this is no longer relevant to the new approach.

File: internal/tofu/node_resource_plan.go (hunk 3 above)
Comment (diofeher) [reply internal/tofu/node_resource_plan.go]: That's good to see! Thanks @Gogotchuri ! Gonna review it

File: internal/tofu/context_import.go [hunk 4]
```diff
@@ -331,6 +343,14 @@ func (c *Context) Import(ctx context.Context, config *configs.Config, prevRunSta
 		return state, diags
 	}
 
+	// Once we have all instances expanded, we are able to do a complete validation for import targets
+	// This part validates imports of both types (import blocks and CLI imports)
+	allInstances := walker.InstanceExpander.AllInstances()
+	importValidateDiags := c.postPlanValidateImports(walker.ImportResolver, allInstances)
```
Comment (diofeher) [inline internal/tofu/context_import.go]: nit: I still think `postExpansionValidateImports` would be a less confusing name for this function, since its logic depends on the expansion rather than the plan itself.

File: internal/tofu/context_import.go [hunk 5]
```diff
@@ -277,6 +277,18 @@ func (ri *ImportResolver) GetImport(address addrs.AbsResourceInstance) *Evaluate
 	return nil
 }
 
+func (ri *ImportResolver) addCLIImportTarget(importTarget *ImportTarget) {
+	ri.mu.Lock()
+	defer ri.mu.Unlock()
+	importAddress := importTarget.CommandLineImportTarget.Addr
+	ri.imports[importAddress.String()] = EvaluatedConfigImportTarget{
+		// Since this import target originates from the CLI, and we have no config block for it
+		Config: nil,
```
Comment (cam72cam) [inline internal/tofu/context_import.go:290]: Please check references of this field as it will cause panics elsewhere in the code.

File: internal/tofu/context_import.go (hunk 4 above)
Comment (cam72cam) [reply internal/tofu/context_import.go]: How does this interact with `import {}` blocks?  As they are not processed, will it cause postPlanValidateImports to fail?

Comment (diofeher) [reply internal/tofu/context_import.go]: The import blocks are processed in a different part (notice this one is on `context_import.go`): 

First, they're add to the `importResolver` here:

https://github.com/opentofu/opentofu/blob/main/internal/tofu/node_resource_plan.go#L175

And then, after the expansion occurs, they are processed here:

https://github.com/opentofu/opentofu/blob/main/internal/tofu/context_plan.go#L807

File: internal/tofu/context_import.go (hunk 5 above)
Comment (Gogotchuri) [reply internal/tofu/context_import.go:290]: I have checked the usage of this config field, and there are no possible paths to dereference it (with a nil value) in the current implementation. CLI import target handling should never go down those paths where the config value is used.
I could add nil checks in methods referencing this field, which might be good for future-proofing. On the other hand, if CLI import target handling went down the wrong path due to some future changes, we might as well panic.
Let me know what you think would be a better approach.

File: internal/tofu/context_import.go (hunk 4 above)
Comment (Gogotchuri) [reply internal/tofu/context_import.go]: Yeah, I will rename the method. Definitely a more suitable name

File: internal/tofu/context_import.go (hunk 5 above)
Comment (cam72cam) [reply internal/tofu/context_import.go:290]: I think adding a comment to the field in the struct which mentions that it can be null for this path would assuage my concerns.

Comment (Gogotchuri) [reply internal/tofu/context_import.go:290]: A comment already exists stating that this field could be null if the import didn't originate from a config file.
I have extended the comment to add more detailed information about the current usage.
See https://github.com/opentofu/opentofu/pull/3106/commits/7ef54171a7f9f742c54fbb57671582bca38cdcce

File: internal/tofu/context_import.go [hunk 6]
```diff
@@ -277,6 +277,18 @@ func (ri *ImportResolver) GetImport(address addrs.AbsResourceInstance) *Evaluate
 	return nil
 }
 
+func (ri *ImportResolver) addCLIImportTarget(importTarget *ImportTarget) {
+	ri.mu.Lock()
+	defer ri.mu.Unlock()
+	importAddress := importTarget.CommandLineImportTarget.Addr
+	ri.imports[importAddress.String()] = EvaluatedConfigImportTarget{
+		// Since this import target originates from the CLI, and we have no config block for it
```
Comment (Yantrio) [inline internal/tofu/context_import.go:287]: Nitpick: I feel like this is half a setence, how do you feel about fleshing this out completely to explain why we're doing what we're doing?

File: internal/tofu/node_resource_plan.go [hunk 7]
```diff
@@ -165,13 +165,19 @@ func (n *nodeExpandPlannableResource) DynamicExpand(evalCtx EvalContext) (*Graph
...
+		// This was added to correctly validate the existence of the targeted resource instances in case for_each key is used, either on a module or on a resource.
+		if importTarget.IsFromImportCommandLine() {
+			importResolver.addCLIImportTarget(importTarget)
+			continue
+		}
+		// If PreDestroyRefresh is true, we know we are running as part of a refresh plan, immediately before a destroy
 		// plan. In the destroy plan mode, import blocks are not relevant, that's why we skip resolving imports
-		skipImports := importTarget.IsFromImportBlock() && !n.preDestroyRefresh
-		if skipImports {
+		if !n.preDestroyRefresh {
```
Comment (Yantrio) [inline internal/tofu/node_resource_plan.go]: I think I prefer the explicit check for IsFromImportBlock here just incase we support importing from other methods in the future. Just to be a little defensive and avoid possible regressions

Comment (Gogotchuri) [reply internal/tofu/node_resource_plan.go]: Good point, I'll make that change. Thanks!

**Review by cam72cam — APPROVED** [review]

File: internal/tofu/node_resource_plan.go (hunk 7 above)
Comment (Gogotchuri) [reply internal/tofu/node_resource_plan.go]: https://github.com/opentofu/opentofu/pull/3106/commits/741191c85ff375d9e4737af7a7450a5d2106277d

File: internal/tofu/context_import.go (hunk 6 above)
Comment (Gogotchuri) [reply internal/tofu/context_import.go:287]: https://github.com/opentofu/opentofu/pull/3106/commits/741191c85ff375d9e4737af7a7450a5d2106277d
Let me know if there's anything else you would like to add.

**Review by Yantrio — APPROVED** [review]
